cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -0,0 +1,194 @@
+	if err != nil {
+		return err
+	}
+	if !isMounted {
+		fmt.Fprintf(stdout, "/dev/disk/by-label/ubuntu-seed %s\n", filepath.Join(runMnt, "ubuntu-seed"))
+		return nil
+	}
+	// XXX: how do we select a different recover system from the cmdline?
+
+	// 2. (auto) select recovery system for now
+	isMounted, err = osutilIsMounted(filepath.Join(runMnt, "base"))
+	if err != nil {
+		return err
+	}
+	if !isMounted {
+		l, err := filepath.Glob(filepath.Join(runMnt, "/ubuntu-seed/systems/*"))
+		if err != nil {
+			return err
+		}
+		if len(l) == 0 {
+			return fmt.Errorf("cannot find a recovery system")
+		}
+		if len(l) > 1 {
+			return fmt.Errorf("cannot use multiple recovery systems yet")
+		}
+		// load the seed and generate mounts for kernel/base
+		label := filepath.Base(l[0])
+		seedDir := filepath.Dir(filepath.Dir(l[0]))
+		deviceSeed, err := seed.Open(seedDir, label)
+		if err != nil {
+			return err
+		}
+		// load assertions into a temporary database
+		if err := deviceSeed.LoadAssertions(nil, nil); err != nil {
+			return err
+		}
+		perf := timings.New(nil)
+		if err := deviceSeed.LoadMeta(perf); err != nil {
+			return err
+		}
+		// XXX: do we need more cross checks here?
+		for _, essentialSnap := range deviceSeed.EssentialSnaps() {
+			snapf, err := snap.Open(essentialSnap.Path)
+			if err != nil {
+				return err
+			}
+			info, err := snap.ReadInfoFromSnapFile(snapf, essentialSnap.SideInfo)
+			if err != nil {
+				return err
+			}
+			switch info.GetType() {
+			case snap.TypeBase:
+				fmt.Fprintf(stdout, "%s %s\n", essentialSnap.Path, filepath.Join(runMnt, "base"))
+			case snap.TypeKernel:
+				fmt.Fprintf(stdout, "%s %s\n", essentialSnap.Path, filepath.Join(runMnt, "kernel"))
+			}
+		}
+
+		return nil
+	}
+
+	// 3. mount "ubuntu-data" on a tmpfs
+	isMounted, err = osutilIsMounted(filepath.Join(runMnt, "ubuntu-data"))
+	if err != nil {
+		return err
+	}
+	if !isMounted {
+		// XXX: is there a better way?
+		fmt.Fprintf(stdout, "--type=tmpfs tmpfs /run/mnt/ubuntu-data\n")
+		return nil
+	}
+
+	// 4. done, no output, no error indicates to initramfs we are done
+	//    with mounting stuff
+	return nil
+}
+
+func generateMountsModeRecover() error {
+	return fmt.Errorf("recover mode mount generation not implemented yet")
+}
+
+func generateMountsModeRun() error {
+	return fmt.Errorf("run mode mount generation not implemented yet")
+}
+
+func isInstallMode(content []byte) bool {
+	// XXX: deal with whitespace
+	if bytes.Contains(content, []byte("snapd_recovery_mode=install")) {
+		return true
+	}
+	// no snapd_recovery_mode var set -> assume install mode
+	if bytes.Contains(content, []byte("snapd_recovery_mode= ")) || bytes.HasSuffix(content, []byte("snapd_recovery_mode=")) {
+		return true
+	}
+	return false
+}
+
+func isRecoverMode(content []byte) bool {
+	// XXX: deal with whitespace
+	return bytes.Contains(content, []byte("snapd_recovery_mode=recover"))
+}
+
+func isRunMode(content []byte) bool {
+	// XXX: deal with whitespace
```

Comment:
That makes run mode explicit. Should it be the default mode if none is specified?

Reply:
I think being explicit about run mode is what we want. The assumption is that if grubenv is missing/broken and we don't have any information about what mode we should pick we go into install mode to give the user a chance to reinstall or to recover. Unless I miss something of course.

Reply:
Actually I think this will probably be tweaked a bit more once we have something that actually boots. This "snap-bootstrap initramfs-mounts" helper will run in both if we run from a recovery grub or from a normal grub. However in the normal case we probably have no "snapd_recovery_mode" set because it does not make much sense. Conversely when running from the recovery grub we probably never set "snapd_recovery_mode=run" so I guess the final shape of this will be that no "snapd_recovery_mode" at all in the commandline means run-mode. Which I guess I what you wanted to say :) But I guess its ok for now because we don't have run mode just yet (but very soon hopefully!).